v2/src/log_sink_etw.c

```diff
@@ -0,0 +1,497 @@
+
+                break;
+            }
+            else
+            {
+                (void)printf("ETW provider was NOT registered.\r\n");
+                (void)InterlockedExchange(&etw_provider_state, PROVIDER_STATE_NOT_REGISTERED);
+            }
+        }
+    }
+}
+
+// want these aligned to 1 byte (copy/pasted from Tracelogging code)
+__pragma(pack(push, 1))
+typedef struct SELF_DESCRIBED_EVENT_TAG
+{
+    uint8_t _tlgBlobTyp;
+    uint8_t _tlgChannel;
+    uint8_t _tlgLevel;
+    uint8_t _tlgOpcode;
+    uint64_t _tlgKeyword;
+    uint16_t _tlgEvtMetaSize;
+    uint16_t _tlgEvtTag;
+    uint8_t metadata[];
+} SELF_DESCRIBED_EVENT;
+__pragma(pack(pop))
+
+// This function was written with a little bit of reverse engineering of TraceLogging and guidance from 
+// the TraceLogging.h header about the format of the self described events
+static void internal_emit_self_described_event(const char* event_name, uint16_t event_name_length, uint8_t trace_level, const LOG_CONTEXT_PROPERTY_VALUE_PAIR* context_property_value_pairs, uint32_t property_value_count, const char* message, const char* file, const char* func, int32_t line, va_list args)
+{
+    TraceLoggingHProvider const _tlgProv = g_my_component_provider;
+    if (trace_level < _tlgProv->LevelPlus1)
+    {
+        // have a stack allocated buffer where we construct the event metadata and
+        uint8_t _tlgEvent[MAX_SELF_DESCRIBED_EVENT_WITH_METADATA];
+        bool add_properties = true;
+
+        // compute event metadata size
+        // event_name_length includes NULL terminator
+        /* Codes_SRS_LOG_SINK_ETW_01_042: [ log_sink_etw.log_sink_log shall compute the metadata size for the self described event metadata as follows: ]*/
+        /* Codes_SRS_LOG_SINK_ETW_01_043: [ Length of the event name (determined at compile time, excluding zero terminator) + 1. ]*/
+        uint32_t i;
+
+        // alias to the event bytes
+        SELF_DESCRIBED_EVENT* self_described_event = (SELF_DESCRIBED_EVENT*)_tlgEvent;
+
+        uint8_t* pos = &self_described_event->metadata[0];
+
+        // at this point we filled all the event information, now we need to provide the metadata bytes
+        // first one is the event name, followed by metadata for all the fields
+        // copy event name 
+        (void)memcpy(pos, event_name, event_name_length); pos += event_name_length;
+
+        // copy the field metadata (name and in type)
+
+        /* Codes_SRS_LOG_SINK_ETW_01_034: [ log_sink_etw.log_sink_log shall fill the event metadata: ]*/
+
+        /* Codes_SRS_LOG_SINK_ETW_01_035: [ The string content (as field name, excluding zero terminator), followed by one byte with the value TlgInANSISTRING. ]*/
+        /* Codes_SRS_LOG_SINK_ETW_01_044: [ Length of the content field name (determined at compile time, excluding zero terminator) + 1. ]*/
+        (void)memcpy(pos, "content", sizeof("content")); pos += sizeof("content");
+        *pos = TlgInANSISTRING;  pos++;
```

Comment:
>pos [](http://example.com/codeflow?start=33&length=3)

while it is pretty obvious that there's NO CHANCE for pos to escape after 4096 bytes, we still should have some static assert that verifies that metadata size can at least accomodate "content" + "file" + "func"+ ...

Reply:
It might be further helped by rearranging the fields so that "event_name" is the last one?

Reply:
I think I want to address this with having prepopulated structures